test/Microsoft.AspNetCore.Server.KestrelTests/BadHttpRequestTests.cs

```diff
@@ -1,72 +1,119 @@
+        [InlineData("GET / HTTP/1.1 \r")]
+        [InlineData("GET / HTTP/1.1a\r")]
+        [InlineData("GET / HTTP/1.0\n\r")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/3.0\r")]
+        [InlineData("GET / H\r")]
+        [InlineData("GET / HTTP/1.\r")]
+        [InlineData("GET / hello\r")]
+        [InlineData("GET / 8charact\r")]
+        // Missing LF
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task TestBadRequestLines(string request)
         {
             using (var server = new TestServer(context => { return Task.FromResult(0); }))
             {
                 using (var connection = new TestConnection(server.Port))
                 {
-                    var receiveTask = Task.Run(async () =>
-                    {
-                        await connection.Receive(
-                            "HTTP/1.1 400 Bad Request",
-                            "");
-                        await connection.ReceiveStartsWith("Date: ");
-                        await connection.ReceiveForcedEnd(
-                            "Content-Length: 0",
-                            "Server: Kestrel",
-                            "",
-                            "");
-                    });
-
-                    try
-                    {
-                        await connection.SendEnd(request).ConfigureAwait(false);
-                    }
-                    catch (Exception)
-                    {
-                        // TestConnection.SendEnd will start throwing while sending characters
-                        // in cases where the server rejects the request as soon as it
-                        // determines the request line is malformed, even though there
-                        // are more characters following.
-                    }
+                    await connection.SendEnd(request);
+                    await ReceiveBadRequestResponse(connection, request == "GET / HTTP/1.1\r");
+                }
+            }
+        }
 
-                    await receiveTask;
+        [Theory]
+        [InlineData(" ")]
+        [InlineData("GET  ")]
+        [InlineData("GET / HTTP/1.2\r")]
+        [InlineData("GET / HTTP/1.0\rA")]
+        public async Task ServerClosesConnectionAsSoonAsBadRequestLineIsDetected(string request)
+        {
+            using (var server = new TestServer(context => { return Task.FromResult(0); }))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.Send(request);
+                    await ReceiveBadRequestResponse(connection, false);
                 }
             }
         }
+
+        private async Task ReceiveBadRequestResponse(TestConnection connection, bool expectedConnectionCloseHeader)
+        {
+            await connection.Receive(
+                "HTTP/1.1 400 Bad Request",
+                "");
+
+            if (expectedConnectionCloseHeader)
```

Comment:
Bad requests should terminate the connection as there maybe more data in the input at the time the request is detected as being bad; which means where the next request begins cannot be determined.